go-controller/pkg/clustermanager/zone_cluster_controller.go

```diff
@@ -0,0 +1,355 @@
+//   - Starts watching the kubernetes nodes
+func (zcc *zoneClusterController) Start(ctx context.Context) error {
+	nodeHandler, err := zcc.retryNodes.WatchResource()
+
+	if err != nil {
+		return fmt.Errorf("unable to watch nodes: %w", err)
+	}
+
+	zcc.nodeHandler = nodeHandler
+	return err
+}
+
+func (zcc *zoneClusterController) Stop() {
+	close(zcc.stopChan)
+	zcc.wg.Wait()
+
+	if zcc.nodeHandler != nil {
+		zcc.watchFactory.RemoveNodeHandler(zcc.nodeHandler)
+	}
+}
+
+// handleAddUpdateNodeEvent handles the add or update node event
+func (zcc *zoneClusterController) handleAddUpdateNodeEvent(node *corev1.Node) error {
+	allocatedNodeId, err := zcc.idAllocator.AllocateId(node.Name)
+	if err != nil {
+		return err
+	}
+
+	if allocatedNodeId == util.GetNodeId(node) {
+		// Nothing to update.
+		return nil
+	}
+
+	klog.V(5).Infof("Allocated id [%d] for node %q", allocatedNodeId, node.Name)
+	return zcc.updateNodeAnnotationWithRetry(node.Name, allocatedNodeId)
+}
+
+// handleAddUpdateNodeEvent handles the delete node event
+func (zcc *zoneClusterController) handleDeleteNode(node *corev1.Node) error {
+	zcc.idAllocator.ReleaseId(node.Name)
+	return nil
+}
+
+func (zcc *zoneClusterController) syncNodes(nodes []interface{}) error {
+	return zcc.syncNodeIds(nodes)
+}
+
+func (zcc *zoneClusterController) syncNodeIds(nodes []interface{}) error {
+	duplicateIdNodes := []string{}
+
+	for _, tmp := range nodes {
+		node, ok := tmp.(*corev1.Node)
+		if !ok {
+			return fmt.Errorf("spurious object in syncNodes: %v", tmp)
+		}
+
+		nodeId := util.GetNodeId(node)
+		if nodeId != -1 {
+			klog.V(5).Infof("Node %s has the id %d set", node.Name, nodeId)
+			if err := zcc.idAllocator.ReserveId(node.Name, nodeId); err != nil {
+				// The id set on this node is duplicate.
+				klog.V(5).Infof("Node %s has a duplicate id [%d] set", node.Name, nodeId)
+				duplicateIdNodes = append(duplicateIdNodes, node.Name)
+			}
+		}
+	}
+
+	for i := range duplicateIdNodes {
+		newNodeId, err := zcc.idAllocator.AllocateId(duplicateIdNodes[i])
+		if err != nil {
+			return fmt.Errorf("failed to allocate id for node %s : err - %v", duplicateIdNodes[i], err)
+		} else {
+			klog.V(5).Infof("Allocated new id [%d] for node %q", newNodeId, duplicateIdNodes[i])
+		}
+	}
+
+	return nil
+}
+
+func (zcc *zoneClusterController) updateNodeAnnotationWithRetry(nodeName string, nodeId int) error {
+	// Retry if it fails because of potential conflict which is transient. Return error in the
+	// case of other errors (say temporary API server down), and it will be taken care of by the
+	// retry mechanism.
+	resultErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		// Informer cache should not be mutated, so get a copy of the object
+		node, err := zcc.watchFactory.GetNode(nodeName)
+		if err != nil {
+			return err
+		}
+
+		cnode := node.DeepCopy()
+		cnode.Annotations = util.UpdateNodeIdAnnotation(cnode.Annotations, nodeId)
+		return zcc.kube.UpdateNode(cnode)
```

Comment:
I'm not sure.  In the previous cluster manager PR I think I got a comment to use Update instead of Patch.

If you think its ok let me know and I'll change it.

Reply:
ack let me ask others on my team for more info on what's the best practice and come back to this comment.

Reply:
ok so this is very funny, I checked the history of events and it seems that was not totally necessary in lieu of the changes you did in #3127 ? maybe @npinaeva has more context here and can correct me if I am wrong...
She did https://github.com/ovn-org/ovn-kubernetes/commit/55a4ea4998aab77931c99f87b48884050e8447cf which I think is what @numansiddique you are referring to to use update instead of patch? -> if I look at current master code things have changed already because in https://github.com/ovn-org/ovn-kubernetes/pull/3127/commits/55a4ea4998aab77931c99f87b48884050e8447cf we removed all that! lol So IMHO only one person is updating the` nodeHostSubnetAllocation` currently right? or is it multiple controllers?

Reply:
Ack.  I had kept this comment to  be resolved at the end as I was not sure how to convert a annotation of type map[string]string to map[string]interface{}

I figured it out now.